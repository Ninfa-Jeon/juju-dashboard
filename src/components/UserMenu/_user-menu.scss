@import "../../scss/settings";
@import "vanilla-framework/scss/vanilla";

.user-menu {
  @include vf-animation(all, brisk, ease-in-out);

  bottom: 0;
  color: $color-light;
  max-height: 2.5rem;
  padding: 0;
  position: absolute;
  width: 100%;

  &.is-active {
    max-height: 120px;
  }

  @at-root {
    .has-collapsible-sidebar .is-collapsed .user-menu.is-active {
      max-height: 2.5rem;
    }
  }

  &__header {
    cursor: pointer;
    display: flex;
    padding: 0 1rem;

    [class^="p-icon"] {
      position: relative;
      top: 0.25rem;
    }

    .p-icon--chevron-up {
      @at-root {
        .is-active .p-icon--chevron-up {
          transform: none;
        }
      }

      transform: rotate(180deg);
    }
  }

  &__name {
    display: inline-block;
    flex-grow: 1;
    padding: 0 1rem;
  }

  &__options {
    margin: 1rem 0;
    padding: 0;
  }

  &__link {
    display: block;
    padding: 0.25rem 1rem;

    &,
    &:visited {
      color: $color-light;
    }

    &:hover {
      background-color: $color-navigation-highlight;
    }
  }
}
