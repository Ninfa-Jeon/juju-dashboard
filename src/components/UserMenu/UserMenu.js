import React, { useEffect } from "react";
import { useSelector, useDispatch, useStore } from "react-redux";
import { Link, NavLink } from "react-router-dom";
import classNames from "classnames";
import { getActiveUserTag, getWSControllerURL } from "app/selectors";

import { logOut } from "app/actions";
import useAnalytics from "hooks/useAnalytics";
import { extractOwnerName } from "app/utils";

import { userMenuActive } from "ui/actions";
import { isUserMenuActive } from "ui/selectors";

import "./_user-menu.scss";

const UserMenu = () => {
  const sendAnalytics = useAnalytics();
  const dispatch = useDispatch();
  const getState = useStore().getState;
  const activeUser = getActiveUserTag(
    useSelector(getWSControllerURL),
    getState()
  );
  const isActive = useSelector(isUserMenuActive) || false;

  useEffect(() => {
    if (isActive) {
      sendAnalytics({
        category: "User",
        action: "Opened user menu",
      });
    }
  }, [isActive, sendAnalytics]);

  const toggleUserMenuActive = () => dispatch(userMenuActive(!isActive));

  return (
    <div
      className={classNames("user-menu", {
        "is-active": isActive,
      })}
    >
      <>
        <div
          className="user-menu__header"
          onClick={toggleUserMenuActive}
          onKeyPress={toggleUserMenuActive}
          role="button"
          tabIndex="0"
        >
          <i className="p-icon--user"></i>
          <span className="user-menu__name">
            {activeUser ? extractOwnerName(activeUser) : ""}
          </span>
          <i className="p-icon--contextual-menu"></i>
        </div>
        <ul className="p-list user-menu__options">
          <li className="p-list__item">
            <NavLink
              className="user-menu__link p-list__link"
              isActive={(match) => {
                if (match && match.url.includes("settings")) {
                  return true;
                }
              }}
              to="/settings"
              activeClassName="is-selected"
            >
              Settings
            </NavLink>
          </li>
          <li className="p-list__item">
            <Link
              className="user-menu__link"
              to="/"
              onClick={() => dispatch(logOut(getState))}
            >
              Log out
            </Link>
          </li>
        </ul>
      </>
    </div>
  );
};

export default UserMenu;
