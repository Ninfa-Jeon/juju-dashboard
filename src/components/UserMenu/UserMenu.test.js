import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { mount } from "enzyme";
import cloneDeep from "clone-deep";
import dataDump from "testing/complete-redux-store-dump";

import UserMenu from "./UserMenu";

const mockStore = configureStore([]);
describe("User Menu", () => {
  it("renders without crashing and matches snapshot", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <UserMenu />
        </Router>
      </Provider>
    );
    expect(wrapper.find(".user-menu")).toMatchSnapshot();
  });

  it("is inactive by default", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <UserMenu />
        </Router>
      </Provider>
    );
    expect(wrapper.find(".user-menu").hasClass("is-active")).toEqual(false);
  });

  it("is active when userMenuActive in redux store is true", () => {
    const clonedDump = cloneDeep(dataDump);
    clonedDump.ui.userMenuActive = true;
    const store = mockStore(clonedDump);
    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <UserMenu />
        </Router>
      </Provider>
    );

    expect(wrapper.find(".user-menu").hasClass("is-active")).toEqual(true);
  });

  it("displays current logged in user", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <UserMenu />
        </Router>
      </Provider>
    );
    const username = ".user-menu__name";
    expect(wrapper.find(username).text()).toEqual("activedev");
  });

  it("Test dispatch function is fired", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <UserMenu />
        </Router>
      </Provider>
    );
    wrapper.find(".user-menu__header").simulate("click");
    const actions = store.getActions();
    const expectedPayload = { payload: true, type: "TOGGLE_USER_MENU" };
    expect(actions).toEqual([expectedPayload]);
  });
});
