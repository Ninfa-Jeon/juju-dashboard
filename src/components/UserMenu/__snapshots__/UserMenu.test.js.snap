// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Menu renders without crashing and matches snapshot 1`] = `
<div
  className="user-menu"
>
  <div
    className="user-menu__header"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <i
      className="p-icon--user"
    />
    <span
      className="user-menu__name"
    >
      activedev
    </span>
    <i
      className="p-icon--contextual-menu"
    />
  </div>
  <ul
    className="p-list user-menu__options"
  >
    <li
      className="p-list__item"
    >
      <NavLink
        activeClassName="is-selected"
        className="user-menu__link p-list__link"
        isActive={[Function]}
        to="/settings"
      >
        <Link
          aria-current={null}
          className="user-menu__link p-list__link"
          to={
            Object {
              "hash": "",
              "pathname": "/settings",
              "search": "",
              "state": null,
            }
          }
        >
          <LinkAnchor
            aria-current={null}
            className="user-menu__link p-list__link"
            href="/settings"
            navigate={[Function]}
          >
            <a
              aria-current={null}
              className="user-menu__link p-list__link"
              href="/settings"
              onClick={[Function]}
            >
              Settings
            </a>
          </LinkAnchor>
        </Link>
      </NavLink>
    </li>
    <li
      className="p-list__item"
    >
      <Link
        className="user-menu__link"
        onClick={[Function]}
        to="/"
      >
        <LinkAnchor
          className="user-menu__link"
          href="/"
          navigate={[Function]}
          onClick={[Function]}
        >
          <a
            className="user-menu__link"
            href="/"
            onClick={[Function]}
          >
            Log out
          </a>
        </LinkAnchor>
      </Link>
    </li>
  </ul>
</div>
`;
